src/forward/compiler.jl

```diff
@@ -70,7 +70,91 @@ function dual(ir)
   return ir
 end
 
-@dynamo function _pushforward(_, x...)
+const chainrules_frule_fallback = which(frule, Tuple{Any})
+
+"""
+  has_chain_frule(T)
+
+For a type-tuple `T` e.g. `Tuple{typeof(f), Int, Float64}`, checks if there is a `rrule` defined for it.
+Excluding the generic fallback.
+The first return value is `true` if the `rrule` exists, `false` otherwise.
+If it does not, then the second argument is a list of edges to attach to the CodeInfo for a generated function,
+such that if a suitable rule is defined later, the generated function will recompile.
+"""
+function has_chain_frule(T)
+  m = meta(Tuple{typeof(frule), T.parameters...})
+
+  if m.method !== chainrules_frule_fallback
+    # found a frule, no need to add any edges
+    return true, nothing
+  end
+
+  return false, m.instance
+end
+
+"""
+    chain_frule(f, args...)
```

Comment:
this doc is wrong but the code is right.
`dargs` includes `dself, darg1, darg2 ...` and `args` includes `foo, darg1, darg2`